rclcpp/include/rclcpp/strategies/static_memory_strategy.hpp

```diff
@@ -0,0 +1,159 @@
+    _pool_seq = 0;
+    _exec_seq = 0;
+
+    for (size_t i = 0; i < _pool_size; ++i) {
+      _memory_map[_memory_pool[i]] = 0;
+    }
+
+    for (size_t i = 0; i < _max_executables; ++i) {
+      _executable_pool[i] = std::make_shared<executor::AnyExecutable>(executor::AnyExecutable());
+    }
+  }
+
+  void ** borrow_handles(HandleType type, size_t number_of_handles)
+  {
+    switch (type) {
+      case HandleType::subscriber_handle:
+        return _subscriber_pool;
+      case HandleType::service_handle:
+        return _service_pool;
+      case HandleType::client_handle:
+        return _client_pool;
+      case HandleType::guard_condition_handle:
+        return _guard_condition_pool;
+      default:
+        break;
+    }
+    throw std::runtime_error("Unrecognized enum, could not borrow handle memory.");
+  }
+
+  void return_handles(HandleType type, void ** handles)
+  {
+    switch (type) {
+      case HandleType::subscriber_handle:
+        memset(_subscriber_pool, 0, _max_subscribers);
+        break;
+      case HandleType::service_handle:
+        memset(_service_pool, 0, _max_services);
+        break;
+      case HandleType::client_handle:
+        memset(_client_pool, 0, _max_clients);
+        break;
+      case HandleType::guard_condition_handle:
+        memset(_guard_condition_pool, 0, _max_guard_conditions);
+        break;
+      default:
+        throw std::runtime_error("Unrecognized enum, could not return handle memory.");
+        break;
+    }
+  }
+
+  executor::AnyExecutableSharedPtr instantiate_next_executable()
+  {
+    if (_exec_seq >= _max_executables) {
+      // wrap around
+      _exec_seq = 0;
+    }
+    size_t prev_exec_seq = _exec_seq;
+    ++_exec_seq;
+
+    return _executable_pool[prev_exec_seq];
+  }
+
+  void * rcl_malloc(size_t size)
+  {
+    // Extremely naive static allocation strategy
+    // Keep track of block size at a given pointer
+    if (_pool_seq + size > _pool_size) {
+      // Start at 0
+      _pool_seq = 0;
+    }
+    void * ptr = _memory_pool[_pool_seq];
+    if (_memory_map.count(ptr) == 0) {
+      // We expect to have the state for all blocks pre-mapped into _memory_map
+      throw std::runtime_error("Unexpected pointer in rcl_malloc.");
+    }
+    _memory_map[ptr] = size;
+    size_t prev_pool_seq = _pool_seq;
+    _pool_seq += size;
+    return _memory_pool[prev_pool_seq];
+  }
+
+  void rcl_free(void * ptr)
+  {
+    if (_memory_map.count(ptr) == 0) {
+      // We expect to have the state for all blocks pre-mapped into _memory_map
+      throw std::runtime_error("Unexpected pointer in rcl_free.");
+    }
+
+    memset(ptr, 0, _memory_map[ptr]);
+  }
+
+protected:
+private:
+  static const size_t _pool_size = 1024;
+  static const size_t _max_subscribers = 10;
+  static const size_t _max_services = 5;
+  static const size_t _max_clients = 10;
+  static const size_t _max_guard_conditions = 50;
+  static const size_t _max_executables = 1;
```

Comment:
I was just thinking arguments to the constructor or a function which lets the user change the limits, with the caveat that those functions will do memory allocation.

Not figuring that out right now is ok though.